geoengine/tasks.py

```diff
@@ -0,0 +1,270 @@
+            return False
+
+        return self.status == other.status and self.clean_up == other.clean_up
+
+    def __str__(self):
+        return f"status={self.status.value}, clean_up={self.clean_up}"
+
+
+class FailedTaskStatusInfo(TaskStatusInfo):
+    '''A wrapper for a failed task status with information about the failure'''
+
+    def __init__(self, status, error, clean_up) -> None:
+        super().__init__(status)
+        self.error = error
+        self.clean_up = clean_up
+
+    def __eq__(self, other):
+        '''Check if two task statuses are equal'''
+        if not isinstance(other, self.__class__):
+            return False
+
+        return self.status == other.status and self.error == other.error and self.clean_up == other.clean_up
+
+    def __str__(self):
+        return f"status={self.status.value}, error={self.error}, clean_up={self.clean_up}"
+
+
+def __task_status_from_response(response: Dict[str, str]) -> TaskStatusInfo:
+    '''
+    Parse a http response to a `TaskStatusInfo`
+
+    The task can be one of:
+    RunningTaskStatusInfo, CompletedTaskStatusInfo, AbortedTaskStatusInfo or FailedTaskStatusInfo
+    '''
+
+    if 'status' not in response:
+        raise GeoEngineException(response)
+
+    status = TaskStatus(response['status'])
+
+    if status == TaskStatus.RUNNING:
+        if 'pct_complete' not in response or 'time_estimate' not in response or 'info' not in response:
+            raise GeoEngineException(response)
+        return RunningTaskStatusInfo(status, response['pct_complete'], response['time_estimate'], response['info'])
+    if status == TaskStatus.COMPLETED:
+        if 'info' not in response or 'timeTotal' not in response:
+            raise GeoEngineException(response)
+        return CompletedTaskStatusInfo(status, response['info'], response['timeTotal'])
+    if status == TaskStatus.ABORTED:
+        if 'cleanUp' not in response:
+            raise GeoEngineException(response)
+        return AbortedTaskStatusInfo(status, response['cleanUp'])
+    if status == TaskStatus.FAILED:
+        if 'error' not in response or 'cleanUp' not in response:
+            raise GeoEngineException(response)
+        return FailedTaskStatusInfo(status, response['error'], response['cleanUp'])
+    raise GeoEngineException(response)
+
+
+class TaskStatusWithId:
+    '''A wrapper for a task id with a task status type'''
+
+    def __init__(self, task_id, task_status_with_info) -> None:
+        self.task_id = task_id
+        self.task_status_with_info = task_status_with_info
+
+    def __eq__(self, other):
+        '''Check if two task ids and task statuses are equal'''
+        if not isinstance(other, self.__class__):
+            return False
+
+        return self.task_id == other.task_id and self.task_status_with_info == other.task_status_with_info
+
+    def __str__(self):
+        return f"TaskId={self.task_id}, TaskInfo={{{self.task_status_with_info}}}"
```

Comment:
You should implement `__repr__`, because then you can just put the status into the notebook or REPL and it outputs a nice representation. For Jupyter, I even found `_repr_html_` as a default method.

Reply:
e3a1a0d536300b261c9d3f90db84a87ac87c08b4